### PR TITLE
Effects: change animated to animationSpeed

### PR DESCRIPTION
**Pull request by Quarkstar**
### Change `animated` to `animationSpeed`

- [x] crt
- [x] glitch
- [x] godray
- [x] oldFilm
- [x] reflection

### Problems
According to their animation, I divide them into 2 groups.
- Depend on `time` (`reflection`, `godray`, `crt`)
  Their animation depend on `filter.time`, so `Animation Speed` has no limits.
- Depend on `seed` (`glitch`, `oldFilm`, `crt`)
  Their animation depend on how often `filter.seed` change(frequency of calling `update`), so they have a max value of `Animation Speed`.
   ```javascript
  update: function(filter, layer) {
     if (filter.animationSpeed !== 0) {
       filter._animationTimer += filter.animationSpeed;
       if (filter._animationTimer >= 1) {
         filter.seed = Math.random();
         filter._animationTimer = 0
       }
     }
  },
  ```
Is that acceptable? I think users may be confused with this. Or is there a way to call `update` once and render it twice?
 
I think `Slider` may be helpful(just between `minSpeed` and `maxSpeed`), or we can keep `animated`.

### CODE REVIEW

**Comment by 4ian** [comment]
> Their animation depend on how often filter.seed change(frequency of calling update), so they have a max value of Animation Speed.

That's ok :) 
As "seed" is a discrete change, i.e: it depends on the frequency of calling update, it can not be animated faster than the number of frames we can display. This is not a problem at all.

> Or is there a way to call update once and render it twice?

That would not work, rendering a frame takes time and this would reduce the framerate because your computer would have more work to do, and so there would be even less frames to be rendered on screen.

So no worries, by definition something that is animated at every update can not go faster than the update refresh rate. It's not an animation but something that is changed randomly, so you can't even notice that things are not as fast as they should.

**Comment by 4ian** [comment]
This being said, your unit is currently tied to the game refresh rate. Instead, we can use a frequency:

```
update: function(filter, layer) {
    if (filter.animationFrequency !== 0) { 
      filter._animationTimer += layer.getElapsedTime() / 1000;
      if (filter._animationTimer >= 1 / filter.animationFrequency) {
        filter.seed = Math.random();
        filter._animationTimer = 0
      }
  }
},
```

`animationFrequency` is in "hz", or if you prefer "number of update per seconds". 

You can also use the approach done in sprites, i.e: ask the user for the time between two updates:

```
update: function(filter, layer) {
    if (filter.animationTimeBetweenUpdates !== 0) { 
      filter._animationTimer += layer.getElapsedTime();
      if (filter._animationTimer >= filter.animationTimeBetweenUpdates) {
        filter.seed = Math.random();
        filter._animationTimer = 0
      }
  }
},
```

`animationTimeBetweenUpdates` is in seconds, or if you prefer "seconds between each update".

I think one of these approach is a bit better than a 0-1 slider for the animation speed. Why? Because if your game runs at 60fps and I put 0.5, the filter would be updated 30 times a second. But if I have a high refresh rate monitor, my game can run at 90fps, and so the filter would be updated 45 times a second. We must at all cost avoid this behavior where a game is running at a different speed according to the monitor.

I would suggest switching to use frequency instead - seems like an intuitive unit (higher frequency = "faster", 0 = "no updates").

**Comment by Quarkstar** [comment]
Thank you! I learned a lot from your advise:)
I have changed all these five filters to `animationSpeed`. And I will upload the gif soon.

**Comment by blurymind** [comment]
Why don't we keep it as it was before, but use what was a boolean as a multiplier.
0 --> disabled
0.5 --> the same calculation as before multiplied by 0.5 (making the speed half)
2.0 --> same as above and so on..

**Comment by Quarkstar** [comment]
I am sorry.
 
But animation of `oldFilmFilter` and `glitchFilter` only depend on `filter.seed`.
```javascript
if ( filter.animationSpeed  > 0 ) {
    filter.time += (layer.getElapsedTime() / 1000) * filter.animationSpeed ;	  
    filter.seed = Math.random();
}
```
This doesn't work on them. When `update` is called, `filter.seed` is updated. So the speed of animation is always the same, and `animationSpeed` doesn't work. That's why I add a `filter._animationTimer`.

I'm sorry, and I'm thinking of a better way to solve it.

**Comment by 4ian** [comment]
> Thats why most video editing/animation software lets you enter FPS and not animationTimeBetweenUpdates

That's why I suggested to use frequency and @Quarkstar updated the PR to use frequency :) 

> Why don't we keep it as it was before - exact same calculation, but use what was an enable boolean as a multiplier instead.

Because this is dependent on the game refresh rate which is tied to the monitor/gpu refresh rate. 

> Remember that an artist is likely entering those values, not a programmer who is obsessed with knowing the exact value between frames :p

That's why frequency is the best thing to use :)

**Comment by 4ian** [comment]
@blurymind See my explanations in https://github.com/4ian/GDevelop/pull/1554#issuecomment-601209066 :)

**Comment by 4ian** [comment]
Note that the frequency is only for `filter.seed`. For animation using `filter.time`, using a multiplier like @blurymind is what we should probably do.

Using a frequency here is **not** applicable. I was only commenting about `filter.seed` updates, **not** `filter.time` - which must be updated continuously.

File: Extensions/Effects/crt-pixi-filter.js [hunk 1]
```diff
@@ -1,13 +1,17 @@
...
   },
   update: function(filter, layer) {
-    if (filter.animated) {
-      filter.time += layer.getElapsedTime() / 1000;
-      filter.seed = Math.random();
+    if (filter.animationFrequency !== 0) { 
+      filter._animationTimer += layer.getElapsedTime() / 1000;
+      if (filter._animationTimer >= 1 / filter.animationFrequency) {
+        filter.seed = Math.random();
+        filter.time += layer.getElapsedTime() / 1000;
```
Comment (4ian) [inline Extensions/Effects/crt-pixi-filter.js]: time must be updated continuously. this is not linked to the frequency, which is only for discrete changes like seed.

File: Extensions/Effects/godray-pixi-filter.js [hunk 2]
```diff
@@ -1,12 +1,16 @@
...
+    godrayFilter._animationTimer = 0;
     return godrayFilter;
   },
   update: function(filter, layer) {
-    if (filter.animated) {
-      filter.time += layer.getElapsedTime() / 1000;
+    if (filter.animationFrequency !== 0) { 
+      filter._animationTimer += layer.getElapsedTime() / 1000;
+      if (filter._animationTimer >= 1 / filter.animationFrequency) {
+        filter.time += layer.getElapsedTime() / 1000;
```
Comment (4ian) [inline Extensions/Effects/godray-pixi-filter.js]: Same here, time must be updated continuously. 
Revert the changes and consider using animationSpeed for this.

File: Extensions/Effects/reflection-pixi-filter.js [hunk 3]
```diff
@@ -19,12 +19,16 @@ gdjs.PixiFiltersTools.registerFilterCreator("Reflection", {
...
+    reflectionFilter._animationTimer = 0;
     return reflectionFilter;
   },
   update: function(filter, layer) {
-    if (filter.animated) {
-      filter.time -= layer.getElapsedTime() / 1000;
+    if (filter.animationFrequency !== 0) { 
+      filter._animationTimer += layer.getElapsedTime() / 1000;
+      if (filter._animationTimer >= 1 / filter.animationFrequency) {
+        filter.time += layer.getElapsedTime() / 1000;
```
Comment (4ian) [inline Extensions/Effects/reflection-pixi-filter.js]: Same, time must be continuously updated. Here you both use a timer and then update the time again with elapsedTime => this does not make sense in terms of time.

Frequency is only for seed.

File: Extensions/Effects/crt-pixi-filter.js (hunk 1 above)
Comment (Quarkstar) [reply Extensions/Effects/crt-pixi-filter.js]: Thank you! I have some misunderstandings about `time` before and I have fixed it.

File: Extensions/Effects/JsExtension.js [hunk 4]
```diff
@@ -887,7 +898,7 @@ module.exports = {
       new gd.PropertyDescriptor(/* defaultValue= */ '1.0')
         .setLabel(_('Vignetting Alpha (between 0 and 1)'))
         .setType('number')
-        .setDescription('The size of the noise particles')
+        .setDescription('Amount of opacity of vignette')
```
Comment (4ian) [inline Extensions/Effects/JsExtension.js]: Nice catch :)

**Comment by 4ian** [comment]
The changes looks correct to me :) I'll let @blurymind have a look - I believe the proper filters have been updated with the proper frequency/animation speed properties but better double check.

We might want to have in another PR special controls shown, like a frequency counter with a slider for frequency, and maybe something similar for the animation speed with a `x` at the beginning or the end of the input (so that it's displaying 0.5x, 1x, 2x..., could help to make more obvious this is a speed multiplier). 
The interesting thing about adding these controls is that they could be used in behaviors or any other thing having properties.

File: Extensions/Effects/crt-pixi-filter.js [hunk 5]
```diff
@@ -5,8 +5,8 @@ gdjs.PixiFiltersTools.registerFilterCreator('CRT', {
     return crtFilter;
   },
   update: function(filter, layer) {
-    if (filter.animated) {
-      filter.time += layer.getElapsedTime() / 1000;
+    if (filter.animationSpeed !== 0) { 
+      filter.time += layer.getElapsedTime() / 1000 * filter.animationSpeed;
       filter.seed = Math.random();
```
Comment (blurymind) [inline Extensions/Effects/crt-pixi-filter.js]: the animation speed multiplier approach doesnt seem to make a difference for this effect. I wonder if we need limit the frequency at which seed is being changed to to it

Comment (4ian) [reply Extensions/Effects/crt-pixi-filter.js]: Good catch! seed and time must be handled separately I think. One with frequency, the other with the animation speed.
So this means two settings in the UI :) 

If that does not make sense to have them separated, then we must update the seed every once in a while, but that means choosing a frequency for the user so seems weird.... It would be great to know what is the difference between time and seed for this CRT filter :)

Comment (Quarkstar) [reply Extensions/Effects/crt-pixi-filter.js]: It works but not obvious. It's much more obvious when `lineWidth` is 10.

The animation of `crt` consists of `noise`(influenced by `seed`) and `line moves`(influenced by `time`).

Comment (blurymind) [reply Extensions/Effects/crt-pixi-filter.js]: I came up with this formula
filter.seed =  Math.round((filter.time % filter.animationSpeed) * 10) + 0.1;

seems to do the trick :)

File: Extensions/Effects/glitch-pixi-filter.js [hunk 6]
```diff
@@ -1,13 +1,16 @@
...
     return glitchFilter;
   },
   update: function(filter, layer) {
-    if (filter.animated) {
-      filter.time += layer.getElapsedTime() / 1000;
-      filter.seed = Math.random();
+    if (filter.animationFrequency !== 0) { 
+      filter._animationTimer += layer.getElapsedTime() / 1000;
+      if (filter._animationTimer >= 1 / filter.animationFrequency) {
+        filter.seed = Math.random();
```
Comment (blurymind) [inline Extensions/Effects/glitch-pixi-filter.js]: If my seed affected by speed formula is ok by @4ian, please simplify this logic to how it was before and use it here too

File: Extensions/Effects/old-film-pixi-filter.js [hunk 7]
```diff
@@ -1,13 +1,16 @@
...
     return oldFilmFilter;
   },  
   update: function(filter, layer) {
-    if (filter.animated) {
-      filter.time += layer.getElapsedTime() / 1000;
-      filter.seed = Math.random();
+    if (filter.animationFrequency !== 0) { 
+      filter._animationTimer += layer.getElapsedTime() / 1000;
+      if (filter._animationTimer >= 1 / filter.animationFrequency) {
+        filter.seed = Math.random();
```
Comment (blurymind) [inline Extensions/Effects/old-film-pixi-filter.js]: again, could be simplified with my seed formula from earlier

File: Extensions/Effects/crt-pixi-filter.js (hunk 5 above)
Comment (Quarkstar) [reply Extensions/Effects/crt-pixi-filter.js]: > If we can give the user control over the speed without adding complexity, that would be great
I do agree with it. But this is a problem with this formula(`oldFilmFilter` as an example), scratches used to be appeared randomly, but now it always appear in the same position.

Comment (blurymind) [reply Extensions/Effects/crt-pixi-filter.js]: > > If we can give the user control over the speed without adding complexity, that would be great
> 
> I do agree with it. But this is a problem with this formula(`oldFilmFilter` as an example), scratches used to be appeared randomly, but now it always appears in the same position. So I think `random` cannot be removed. What do you think?

I think the frequency is kind of counterintuitive, since a greater value results in slower animation. I think we can improve it to make it consistent :)

Comment (blurymind) [reply Extensions/Effects/crt-pixi-filter.js]: put this in old film filter- exact same logic, but with your  renamed parameters
```
update: function(filter, layer) {
    if (filter.animationFrequency !== 0) {
      filter._animationTimer +=
        (layer.getElapsedTime() / 1000) * filter.animationFrequency;
      filter.seed =
        Math.round((filter._animationTimer % filter.animationFrequency) * 10) +
        0.1;
    }
  },
```
the formula works and a greater value results in faster animation, not the other way around

Comment (4ian) [reply Extensions/Effects/crt-pixi-filter.js]: > Math.round((filter._animationTimer % filter.animationFrequency) * 10) + 0.1;

What are the reason for choosing 10 and 0.1?
This seems to also severly limit the 

> filter._animationTimer += (layer.getElapsedTime() / 1000) * filter.animationFrequency;

A frequency is not the same unit as a multiplier, which does not have a unit 😬
A frequency unit is "per second". We should at least be following proper units or not call this a frequency.

Not saying that it's not a good idea to have the two tied - but here we're getting result that is not predictable for the end user.

> the formula works and a greater value results in faster animation, not the other way around

This is the same with a frequency and an animation speed right?
2x, 3x, 4x => you're increasing the speed of the animation

20hz, 30hz, 60hz => you're increasing the speed of the update to the random noise shown on screen.

Comment (4ian) [reply Extensions/Effects/crt-pixi-filter.js]: > The animation of crt consists of noise(influenced by seed) and line moves(influenced by time).

This seems that the CRT filter has two things: noise (seed) and how line moves (time).
If they are two different things, why not showing the two of them in the UI?

Noise update: 30hz, 60hz, 120hz...
Line movement speed: 0.5x, 1x, 2x...

Sounds easy no? And very flexible :) We're not limiting the effect compared to the original Pixi effect, and we're using units that are the proper one.

Comment (blurymind) [reply Extensions/Effects/crt-pixi-filter.js]: I guess it would be even better to keep the randomness, but also keep it animationSpeed somehow?

Comment (4ian) [reply Extensions/Effects/crt-pixi-filter.js]: Yep! That's why I think we need both, but separate (only in the case of CRT I think):

```
update: function(filter, layer) {
    // Update time (for lines on screen)
    filter.time += layer.getElapsedTime() / 1000 * filter.animationSpeed;

    // Update random scratches (at the specified frequency)
    if (filter.animationFrequency !== 0) { 
      filter._animationTimer += layer.getElapsedTime() / 1000;
      if (filter._animationTimer >= 1 / filter.animationFrequency) {
        filter.seed = Math.random();
        filter._animationTimer = 0;
      }
  }
},
```

Comment (Quarkstar) [reply Extensions/Effects/crt-pixi-filter.js]: Thank you! And I have fixed it.

Comment (blurymind) [reply Extensions/Effects/crt-pixi-filter.js]: I like the idea of splitting the random seed from time and letting the user set them individually. If this approach also means that for both a higher value results in a faster animation, then I'm good with it :) one thing worth double checking is if the default values that are set when adding each filter best demonstrate it's animation

Comment (blurymind) [reply Extensions/Effects/crt-pixi-filter.js]: Also check if changing each results in a visually different result. I know it's kind of obvious, but worth double checking

Comment (Quarkstar) [reply Extensions/Effects/crt-pixi-filter.js]: > Also check if changing each results in a visually different result. I know it's kind of obvious, but worth double checking

I have checked them and no problems.

>if the default values that are set when adding each filter best demonstrate it's animation

The default value of `lineWidth`, `lineContrast`, `noiseSize` in `CRT` may need to change. But I don't know which value is appropriate because it looks different in different game. 
I have tried it and I think `lineWidth: 3, lineContrast: 0.2, noiseSize: 2` looks fine to me. What do you think?

Comment (4ian) [reply Extensions/Effects/crt-pixi-filter.js]: > I have tried it and I think lineWidth: 3, lineContrast: 0.2, noiseSize: 2 looks fine to me

Add a screenshot so that we can all see the result :) I would say that any default value that make a pretty rendering on screen are fine :)

File: Extensions/Effects/JsExtension.js [hunk 8]
```diff
@@ -613,10 +625,11 @@ module.exports = {
         .setDescription('The resolution of the displacement image')
     );
     glitchProperties.set(
-      'animated',
-      new gd.PropertyDescriptor(/* defaultValue= */ 'true')
-        .setLabel(_('Animated (Enable animations)'))
-        .setType('boolean')
+      'animationFrequency',
+      new gd.PropertyDescriptor(/* defaultValue= */ '0')
```
Comment (4ian) [inline Extensions/Effects/JsExtension.js]: The default value should not be 0 I think

File: Extensions/Effects/JsExtension.js [hunk 9]
```diff
@@ -897,10 +915,11 @@ module.exports = {
         .setDescription('Blur intensity of the vignette')
     );
     oldFilmProperties.set(
-      'animated',
-      new gd.PropertyDescriptor(/* defaultValue= */ 'true')
-        .setLabel(_('Animated (Enable animations)'))
-        .setType('boolean')
+      'animationFrequency',
+      new gd.PropertyDescriptor(/* defaultValue= */ '0')
```
Comment (4ian) [inline Extensions/Effects/JsExtension.js]: Same, let's make a default value that was equivalent to animated = true, i.e: 60

File: Extensions/Effects/crt-pixi-filter.js (hunk 5 above)
Comment (Quarkstar) [reply Extensions/Effects/crt-pixi-filter.js]: There are three examples.
### Pixi-filter documentation
Screenshot from 2020-03-22 20-40-45

### Default value we used to use
Screenshot from 2020-03-22 20-27-21

### Pixi-filter Demo
Screenshot from 2020-03-22 20-40-45

It is hard to say which looks best to me. Can you choose one? And I will both fix it and the default frequency.

File: Extensions/Effects/JsExtension.js (hunk 8 above)
Comment (blurymind) [reply Extensions/Effects/JsExtension.js]: yes, for every animated filter the default should be for it to animate

File: Extensions/Effects/crt-pixi-filter.js (hunk 5 above)
Comment (4ian) [reply Extensions/Effects/crt-pixi-filter.js]: I'm fine with either pixi-filter doc/demo :)

File: Extensions/Effects/JsExtension.js (hunk 8 above)
Comment (Quarkstar) [reply Extensions/Effects/JsExtension.js]: Thank you! I have fixed it.

And I move animation of `crt-lines` 10 times faster to make it obvious.
Sorry for no gif. (I cannot upload a `gif` which can show the animation clearly and <= 10M).

File: Extensions/Effects/JsExtension.js [hunk 10]
```diff
@@ -434,10 +434,22 @@ module.exports = {
...
+        .setDescription(
+          _(
+            '(E.g., 0: Pause, 0.5: Half speed, 1: Normal speed, 2: Double speed)'
+          )
+        )
+    );
+    crtProperties.set(
+      'animationFrequency',
+      new gd.PropertyDescriptor(/* defaultValue= */ '60')
+        .setLabel(_('Animation Frequency of Noise'))
```
Comment (4ian) [inline Extensions/Effects/JsExtension.js]: Frequency of noise

File: Extensions/Effects/JsExtension.js [hunk 11]
```diff
@@ -887,7 +905,7 @@ module.exports = {
       new gd.PropertyDescriptor(/* defaultValue= */ '1.0')
         .setLabel(_('Vignetting Alpha (between 0 and 1)'))
         .setType('number')
-        .setDescription('The size of the noise particles')
+        .setDescription('Amount of opacity of vignette')
```
Comment (4ian) [inline Extensions/Effects/JsExtension.js]: Actually we can remove the setDescription entirely, this is not giving much information.

File: Extensions/Effects/crt-pixi-filter.js [hunk 12]
```diff
@@ -1,13 +1,19 @@
...
-
+    crtFilter._animationTimer = 0;
     return crtFilter;
   },
   update: function(filter, layer) {
-    if (filter.animated) {
-      filter.time += layer.getElapsedTime() / 1000;
-      filter.seed = Math.random();
+    if (filter.animationSpeed !== 0) { 
+      filter.time += layer.getElapsedTime() / 100 * filter.animationSpeed;
```
Comment (4ian) [inline Extensions/Effects/crt-pixi-filter.js]: `/ 100` looks like a bug. Avoid magic numbers, prefer doing:
`/ 1000 * 10 * filter.animationSpeed`

and add a comment:
`// Multiply by 10 so that the default value shows use a sensible speed`

File: Extensions/Effects/JsExtension.js (hunk 11 above)
Comment (Quarkstar) [reply Extensions/Effects/JsExtension.js]: Thank you and how about `Vignetting Blur`'s description? I think it can be removed too.

File: Extensions/Effects/JsExtension.js (hunk 8 above)
Comment (Quarkstar) [reply Extensions/Effects/JsExtension.js]: ezgif com-video-to-gif

File: Extensions/Effects/JsExtension.js (hunk 11 above)
Comment (4ian) [reply Extensions/Effects/JsExtension.js]: Yes we can remove it too I think! :)

**Comment by 4ian** [comment]
Alright, I think we're good to go with this. Knowing that this is a "breaking" change because filters that were set up as not animated will need to be manually updated by the users to set animation speed or frequency to 0.
I think it's fine but worth noting.

**Comment by 4ian** [comment]
Merged! Thanks for your work on this :) 
I believe having custom controls for the frequency/speed would be awesome (slider + field), though it's maybe not the highest priority (up to you)

**Comment by Quarkstar** [comment]
Thank you!
Actually, I don't know much about ReactJS and I am learning it now. And I will try it when I am more familiar with it. :rofl:  
Thanks again for your patience!
